File: services/video-assembly/domain/clip_rules.py
"""Luật validate độ dài clip dọc theo preset (CR-007 FR19.6/19.7) — HÀM THUẦN.

Không đọc file, không gọi ffmpeg. Ngưỡng preset là config (C2b của CR-007 —
YouTube/TikTok đổi ngưỡng theo thời gian), đọc từ biến môi trường theo đúng
convention của `domain/qc_rules.py::QCThresholds.from_env()`.
"""

from __future__ import annotations

import os
from dataclasses import dataclass

CLIP_PRESET_SHORT_MAX_SECONDS = 60.0
CLIP_PRESET_LONG_MIN_SECONDS = 60.0
CLIP_PRESET_LONG_MAX_SECONDS = 180.0

PRESET_SHORT = "short"
PRESET_LONG = "long"


def _as_float(raw: str | None, default: float) -> float:
    """Giá trị không parse được rơi về mặc định thay vì ném — cùng lý do
    `qc_rules._as_float`: một biến môi trường gõ sai không được làm sập
    video-assembly."""
    if raw is None or raw.strip() == "":
        return default
    try:
        return float(raw)
    except ValueError:
        return default


@dataclass(frozen=True)
class ClipThresholds:
    """FR19.6/C2b — ngưỡng độ dài mỗi preset, cấu hình được bằng biến môi
    trường vì bên thứ ba (YouTube/TikTok) đổi ngưỡng theo thời gian."""

    short_max_seconds: float = CLIP_PRESET_SHORT_MAX_SECONDS
    long_min_seconds: float = CLIP_PRESET_LONG_MIN_SECONDS
    long_max_seconds: float = CLIP_PRESET_LONG_MAX_SECONDS

    @classmethod
    def from_env(cls, env: dict | None = None) -> "ClipThresholds":
        source = os.environ if env is None else env
        get = source.get
        return cls(
            short_max_seconds=_as_float(
                get("CLIP_PRESET_SHORT_MAX_SECONDS"), CLIP_PRESET_SHORT_MAX_SECONDS
            ),
            long_min_seconds=_as_float(
                get("CLIP_PRESET_LONG_MIN_SECONDS"), CLIP_PRESET_LONG_MIN_SECONDS
            ),
            long_max_seconds=_as_float(
                get("CLIP_PRESET_LONG_MAX_SECONDS"), CLIP_PRESET_LONG_MAX_SECONDS
            ),
        )


def validate_clip_duration(
    duration_seconds: float, preset: str, thresholds: ClipThresholds
) -> str | None:
    """Trả `None` nếu `duration_seconds` hợp lệ cho `preset`, hoặc thông báo
    lỗi rõ ràng nếu không (FR19.6 — không im lặng cắt cụt).

    FR19.7: mỗi preset được validate độc lập — một đoạn không hợp `short`
    vẫn có thể hợp `long`, gọi hàm này riêng cho từng preset và không đánh
    đổ preset kia.
    """
    if preset == PRESET_SHORT:
        if duration_seconds <= 0:
            return f"đoạn {duration_seconds:.1f}s không hợp lệ cho preset short"
        if duration_seconds > thresholds.short_max_seconds:
            return (
                f"đoạn {duration_seconds:.1f}s không đủ ngắn cho preset short "
                f"(tối đa {thresholds.short_max_seconds:.0f}s)"
            )
        return None
    if preset == PRESET_LONG:
        if duration_seconds < thresholds.long_min_seconds:
            return (
                f"đoạn {duration_seconds:.1f}s không đủ dài cho preset long "
                f"({thresholds.long_min_seconds:.0f}-{thresholds.long_max_seconds:.0f}s)"
            )
        if duration_seconds > thresholds.long_max_seconds:
            return (
                f"đoạn {duration_seconds:.1f}s không đủ dài cho preset long "
                f"({thresholds.long_min_seconds:.0f}-{thresholds.long_max_seconds:.0f}s)"
            )
        return None
    return f"preset không xác định: {preset!r}"

File: services/orchestrator/internal/domain/clip_rules.go
package domain

import (
	"fmt"
	"os"
	"strconv"
)

// Preset length thresholds for CR-007's vertical clips (FR19.5/C2b). Mirrors
// video-assembly/domain/clip_rules.py's CLIP_PRESET_* constants exactly —
// both services validate independently (D4: GUI catches it early, video-
// assembly re-checks against the real post-intro offset), so they must agree
// on the same env var names and defaults or a request the GUI accepted could
// still be rejected downstream for a different reason.
const (
	defaultClipPresetShortMaxSeconds = 60.0
	defaultClipPresetLongMinSeconds  = 60.0
	defaultClipPresetLongMaxSeconds  = 180.0

	ClipPresetShort = "short"
	ClipPresetLong  = "long"
)

// clipFloatEnvOrDefault reads a float env var, falling back to def when the
// variable is unset or does not parse — same posture as qc_rules.py's
// _as_float: a mistyped threshold must not crash validation, only leave it
// at its shipped default.
func clipFloatEnvOrDefault(key string, def float64) float64 {
	raw := os.Getenv(key)
	if raw == "" {
		return def
	}
	v, err := strconv.ParseFloat(raw, 64)
	if err != nil {
		return def
	}
	return v
}

// ValidateClipDuration reports whether durationSeconds is within bounds for
// preset (FR19.6), reading thresholds from CLIP_PRESET_SHORT_MAX_SECONDS /
// CLIP_PRESET_LONG_MIN_SECONDS / CLIP_PRESET_LONG_MAX_SECONDS (C2b — third
// parties change these over time, so they must be config, not a constant).
//
// Returns nil when durationSeconds fits preset, or an error with a specific,
// actionable message otherwise — FR19.6 explicitly rules out silently
// truncating a segment that does not fit.
func ValidateClipDuration(durationSeconds float64, preset string) error {
	shortMax := clipFloatEnvOrDefault("CLIP_PRESET_SHORT_MAX_SECONDS", defaultClipPresetShortMaxSeconds)
	longMin := clipFloatEnvOrDefault("CLIP_PRESET_LONG_MIN_SECONDS", defaultClipPresetLongMinSeconds)
	longMax := clipFloatEnvOrDefault("CLIP_PRESET_LONG_MAX_SECONDS", defaultClipPresetLongMaxSeconds)

	switch preset {
	case ClipPresetShort:
		if durationSeconds <= 0 {
			return fmt.Errorf("duration %.1fs is not valid for preset \"short\"", durationSeconds)
		}
		if durationSeconds > shortMax {
			return fmt.Errorf("duration %.1fs is too long for preset \"short\" (max %.0fs)", durationSeconds, shortMax)
		}
		return nil
	case ClipPresetLong:
		if durationSeconds < longMin || durationSeconds > longMax {
			return fmt.Errorf("duration %.1fs is out of range for preset \"long\" (%.0f-%.0fs)", durationSeconds, longMin, longMax)
		}
		return nil
	default:
		return fmt.Errorf("unknown clip preset %q", preset)
	}
}
